### DISCUSSION TITLE
Hardforking evmos to cronos

### DISCUSSION THREAD

**Discussion by zenodeapp** [crypto-org-chain/cronos#1280] [Q&A]
Hi, I was wondering if it's possible to perform a hard fork at a specific height to cronos. I managed to get this to work by setting a specific halt-height in my app.toml file of the evmos binary, installing the cronos binary after it halted and allowing it to continue.

I wonder if this would cause issues with older transactions/account balances etc. I'll test this out soon, though wanted to ask if this approach is somewhat valid since I was looking at the Cronos docs and saw that you guys perform hard forks more often (https://docs.cronos.org/for-node-hosts/running-nodes/cronos-mainnet#step-0-notes-on-huygen-network-upgrade).

And if this approach is okay, could this process also be automated via, let's say, migration/upgrade handlers put in place in the app.go file of the evmos binary?

**The idea is to not jump to a new chain, but rather continue a running evmos forked chain as a cronos fork.**

---

I'm still learning and have nearly scratched the surface of the inner workings of cosmos chains, so it would be great if someone could give some pointers or directions. And if this approach isn't going to work out, how should I tackle the issue?

**Comment by yihuang** [comment]
you just need to handle the state migration in upgrade handler, otherwise, it should work just like the other upgrades(hard-forks).

For how cosmos chains handling upgrades(hard-forks) in general, you can check the doc of [upgrade module](https://github.com/cosmos/cosmos-sdk/tree/main/x/upgrade)

**Reply by zenodeapp** [reply]
Hey thanks for the reply!

I'm trying it atm, slowly upping the version. From evmos (44.5) to v0.6 cronos worked well. Now from evmos (44.5) to v0.7 cronos I stumbled upon one problem. Where the feemarket got added as a module, but we already had that module added.

Now I wonder whether I have to up the module version of feemarket in some way or do I simply remove this part where it gets added?

```go
	if upgradeInfo.Name == plan0_7_0 && !app.UpgradeKeeper.IsSkipHeight(upgradeInfo.Height) {
		storeUpgrades := storetypes.StoreUpgrades{
			Added: []string{"feemarket"},
		}

		// configure store loader that checks if version == upgradeHeight and applies store upgrades
		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
	}
```

_(Removing it works, though I'm not sure if this is the correct way)._


This comes from: https://github.com/crypto-org-chain/cronos/blob/v0.7.0/app/app.go#L735-L742.


I'm first trying it one by one, until I can create a single plan that goes to version v1.0.14 (or am I going at it the wrong way?)

**Reply by zenodeapp** [reply]
Ah I think I'm confusing custom modules vs existing modules.

**Reply by zenodeapp** [reply]
Ah well, ignore this! Already seeing how it migrates from version to version when the upgrade is being applied. I'm at v1.0.2 now.

Now I wonder if I still have to add this:

```go
		// Override feemarket parameters
		initialBaseFee := int64(5000000000000)
		app.FeeMarketKeeper.SetParams(ctx, feemarkettypes.Params{
			NoBaseFee:                false,
			BaseFeeChangeDenominator: 100000000,
			ElasticityMultiplier:     2,
			InitialBaseFee:           				initialBaseFee,
			EnableHeight:             0,
		})
		app.FeeMarketKeeper.SetBaseFee(ctx, big.NewInt(initialBaseFee))
```

In this version InitialBaseFee doesn't exist anymore. Has this been adapted to BaseFee? Or is this not necessary to add in v1.0.2 and up?

**Reply by zenodeapp** [reply]
Again, nvm! 1.0.14 shows me how to change this with the forks.go file 😅!

**Reply by yihuang** [reply]
`forks.go` is a more "primitive" way to do breaking upgrade, you should almost always use the upgrade module and gov proposal.

**Reply by zenodeapp** [reply]
Ah got it! Thank you, I'm learning quite a lot from looking at all the code.

I'll make sure to make the changes to the feemarket (if we actually need to do this) in the upgrade handler instead!

**Reply by zenodeapp** [reply]
Okay, interesting. v1.0.2 worked. But going to v1.0.14 causes a cronos mismatch root store:
```
failed to load latest version: version of store cronos mismatch root store's version; expected 99 got 0
```

Do I have to manually add this, similar to:
```go
storeUpgrades := storetypes.StoreUpgrades{
				Added: []string{ibcfeetypes.StoreKey},
			}
```

Weird that it only now starts to say this.

**Reply by yihuang** [reply]
probably

**Reply by zenodeapp** [reply]
Okay that worked, though weird that this only now happened in this version hehe. Thank you. Now a gas error, but that's probably for forgetting to update the toml files.

Gotta love testing and going back and forth between states huh hehe.

**Reply by zenodeapp** [reply]
Just one final question:
How would I know what I should set as BaseFee, MinGasPrice, Elasticity, etc?

The chain's coin will be in 18 decimals (so: sdk.DefaultPowerReduction = sdk.NewIntFromUint64(1000000000000000000))

```go
		params := app.FeeMarketKeeper.GetParams(ctx)
		params.BaseFeeChangeDenominator = 300
		params.ElasticityMultiplier = 4
		params.BaseFee = sdk.NewInt(10000000000000)
		params.MinGasPrice = sdk.NewDec(10000000000000)
		app.FeeMarketKeeper.SetParams(ctx, params)
```

I'm getting a gas to suggest must be positive error and I got a feeling this is caused by decimals. Cronos was how many decimals?

**Reply by zenodeapp** [reply]
Hmm interesting, this error appeared when I had Rosetta turned on (on v1.0.14). Now it's gone when I turn that off.

https://github.com/cosmos/rosetta/blob/main/config.go#L130-L132.

**Reply by yihuang** [reply]
> Cronos was how many decimals?

cro has 18 decimals on cronos.

**Reply by zenodeapp** [reply]
Hi @yihuang, I've been creating scripts and writing documentations to make the onboarding happen more swiftly. The code itself is somewhat ready, but just need to fix some things (like gas configurations, which I have to discuss more with the community, validators and devs).

I just have a couple questions, which are more so related to making it ready for release and that you perhaps know a lot more about:
- Is it wise to edit the go.mod file and relink everything to our own repo instead of the crypto-org-chain? And how would this work with different branches? (see repo: [`genesis-crypto`](https://github.com/zenodeapp/genesis-crypto) for a bit more context). Would I also have to edit individual files to achieve this?
- The releases you guys make for multiple architectures, do you use the release-dry-run script for this? I'm atm building via that, though not sure if that's the actual approach. Takes quite some time also (the release-dry-run command in the Makefile is what I ran, which let me install nix and also configure this: https://github.com/srid/emanote/issues/300#issuecomment-1152396339). Unsure if I'm going at it the right way. It's still building as I'm typing this, so can't conclude what it's exactly creating.

Trial and error can be fun, but could also take a lot more time and is more prone to errors 😣!

**Reply by zenodeapp** [reply]
Alright hehe, figured both out. Just don't have a Darwin-based system to create such releases. Next prob to deal with :)...
